pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,299 @@
+	klog.V(3).Infof("In federated cluster watcher OnUpdate - OLD OBJ: %#v\nNEW OBJ: %#v", oldObj, newObj)
+
+	oldCluster := oldObj.(*fedv1.KubeFedCluster)
+	newCluster := newObj.(*fedv1.KubeFedCluster)
+
+	if reflect.DeepEqual(oldCluster.Spec, newCluster.Spec) {
+		klog.V(2).Infof("KubeFedClusterSpecs are equal - not updating")
+		return
+	}
+
+	kubeConfig, err := f.buildFederatedClusterConfig(newCluster)
+	if err != nil {
+		klog.Errorf("Error building kube config for federated cluster %#v: %v", newCluster, err)
+		return
+	}
+
+	klog.Infof("KubeFedCluster \"%s\" has been updated", newCluster.Name)
+
+	f.Lock()
+	defer f.Unlock()
+
+	if f.clusterMap[newCluster.Name] == nil {
+		klog.Warningf("OnUpdate - no cached entry for KubeFedCluster \"%s\" exists", newCluster.Name)
+		return
+	}
+
+	f.clusterMap[newCluster.Name] = kubeConfig
+
+	for _, clusterWatcher := range f.clusterWatchers {
+		clusterWatcher.onUpdate(newCluster.Name, kubeConfig)
+	}
+}
+
+func (f *federator) buildFederatedClusterConfig(kubeFedCluster *fedv1.KubeFedCluster) (*rest.Config, error) {
+	if kubeFedCluster.Spec.APIEndpoint == "" {
+		return nil, fmt.Errorf("the API endpoint is empty")
+	}
+
+	secretName := kubeFedCluster.Spec.SecretRef.Name
+	if secretName == "" {
+		return nil, fmt.Errorf("the cluster does not have a secret name")
+	}
+
+	secret, err := f.kubeFedClient.CoreV1().Secrets(kubeFedNamespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting Secret \"%s\": %v", secretName, err)
+	}
+
+	token, tokenFound := secret.Data[tokenKey]
+	if !tokenFound || len(token) == 0 {
+		return nil, fmt.Errorf("the secret %#v is missing a non-empty value for %q", secret, tokenKey)
+	}
+
+	kubeConfig, err := clientcmd.BuildConfigFromFlags(kubeFedCluster.Spec.APIEndpoint, "")
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig.CAData = kubeFedCluster.Spec.CABundle
+	kubeConfig.BearerToken = string(token)
+
+	return kubeConfig, nil
+}
+
+func newKubeFedClusterInformer(kubeFedConfig *rest.Config) (cache.SharedIndexInformer, error) {
+	var obj runtime.Object = &fedv1.KubeFedCluster{}
+
+	gvk, err := apiutil.GVKForObject(obj, scheme.Scheme)
+	if err != nil {
+		return nil, fmt.Errorf("error finding GroupVersionKind for %T: %v", obj, err)
+	}
+
+	klog.V(3).Infof("GroupVersionKind for %T: %#v", obj, gvk)
+
+	mapper, err := apiutil.NewDiscoveryRESTMapper(kubeFedConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapper from config: %v", err)
+	}
+
+	mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapping for %#v: %v", gvk, err)
+	}
+
+	client, err := apiutil.RESTClientForGVK(gvk, kubeFedConfig, scheme.Codecs)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTClient for %#v: %v", gvk, err)
+	}
+
+	klog.V(3).Infof("Got RESTClient: %#v", client)
+
+	return cache.NewSharedIndexInformer(
+		&cache.ListWatch{
```

Comment:
Any reason to create this verbosely instead of using something like cache.NewListWatchFromClient ?

Reply:
No can do with `cache.NewListWatchFromClient` unfortunately. It hardcodes `metav1.ParameterCodec` - we need kubefed's `ParameterCodec` and there's no `NewListWatchFromClient` variant that allows for a custom `ParameterCodec`.  In other words, `cache.NewListWatchFromClient` can only be used with k8s core types.